### PR TITLE
SMTP + Base64 Decoder optimizations

### PR DESCRIPTION
**Pull request by inashivb**
Previous PR: https://github.com/OISF/suricata/pull/7472

Changes since v2:
- Fixed `implicit declaration of function 'DEBUG_VALIDATE_BUG_ON'`

### CODE REVIEW

**Comment by suricata-qa** [comment]
ERROR:
 ERROR: ASAN TEST FAIL in tlpr1_asan_suri QA test

ERROR: QA failed on tlpr1_asan_suri.
  

Pipeline 7680

File: src/app-layer-smtp.c [hunk 1]
```diff
@@ -1277,43 +1278,44 @@ static int SMTPProcessRequest(SMTPState *state, Flow *f,
...
-            consumed_line = lf_idx - state->current_line + 1;
-            if (orig_input_len >= idx && idx > 0 && state->input[idx - 1] == 0x0d) {
-                delim_len = 2;
-            } else if (orig_input_len == 1 ||
-                       (orig_input_len >= idx && idx > 0 && state->input[idx - 1] != 0x0d)) {
-                delim_len = 1;
+            state->current_line_len = current_line_consumed - state->current_line_delimiter_len;
+            state->consumed = total_consumed;
+            state->input_len -= current_line_consumed;
+            BUG_ON(state->consumed + state->input_len != state->orig_input_len);
```
Comment (victorjulien) [inline src/app-layer-smtp.c:1312]: DEBUG_VALIDATE_BUG_ON

File: src/util-decode-mime.c [hunk 2]
```diff
@@ -1311,89 +1330,83 @@ static int ProcessBase64BodyLine(const uint8_t *buf, uint32_t len,
...
+        }
+        if (leftover_bytes < B64_BLOCK) {
+            memcpy(state->bvremain, buf + offset + consumed_bytes, leftover_bytes);
+            state->bvr_len = leftover_bytes;
+            return MIME_DEC_OK;
+        }
+        /* Update counts */
+        remaining -= consumed_bytes;
+        offset += consumed_bytes;
+        /* If remaining is 4 by this time, it's likely due to error/spaces during processing */
```
Comment (victorjulien) [inline src/util-decode-mime.c:1397]: comment needs to be updated to reflect updated error handling

File: src/util-decode-mime.c [hunk 3]
```diff
@@ -1311,89 +1330,83 @@ static int ProcessBase64BodyLine(const uint8_t *buf, uint32_t len,
...
+        if (remaining == 4) {
+            memcpy(state->bvremain, buf + offset, remaining);
+            state->bvr_len = remaining;
+            break;
+        }
+        if ((remaining > 0 && remaining < B64_BLOCK) ||
```
Comment (victorjulien) [inline src/util-decode-mime.c:1403]: nit: `else if`

File: src/util-decode-mime.c [hunk 4]
```diff
@@ -1311,89 +1330,83 @@ static int ProcessBase64BodyLine(const uint8_t *buf, uint32_t len,
...
+
     /* First process remaining from previous line */
     if (state->bvr_len > 0) {
-
-        SCLogDebug("Base64 line remainder found: %u", state->bvr_len);
-
         /* Process remainder and return number of bytes pulled from current buffer */
         rem1 = ProcessBase64Remainder(buf, (uint8_t)len, state, 0);
+        int32_t remainder_b64 = len - rem1;
+        if (remainder_b64 < B64_BLOCK) {
```
Comment (victorjulien) [inline src/util-decode-mime.c:1344]: I see that `remainder_b64` can be 0, which leads to a useless memcpy call. But what does it mean if it is 0?

File: src/util-decode-mime.c [hunk 5]
```diff
@@ -1311,89 +1330,83 @@ static int ProcessBase64BodyLine(const uint8_t *buf, uint32_t len,
                 len, MAX_ENC_LINE_LEN);
     }
 
+    if (state->bvr_len + len < B64_BLOCK) {
```
Comment (victorjulien) [inline src/util-decode-mime.c:1333]: can len be 0 here?

**Comment by victorjulien** [comment]
Replaced by #7482

**Pull request closed without merging**